### PR TITLE
refactor(utils/basic-auth):  Moved Internal function to utils 

### PR DESCRIPTION
**Pull request by sugar-cat7**
I have moved the functions used internally in the Basic Auth Middleware to the utils directory. Additionally, I have added tests.

For the background on this refactor, please refer to the following: https://github.com/honojs/middleware/pull/676#issuecomment-2323178685

### The author should do the following, if applicable

- [ ] Add tests
- [x] Run tests
- [x] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

### CODE REVIEW

File: src/utils/basic-auth.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import type { HonoRequest } from '../request'
+import { decodeBase64 } from './encode'
+
+const CREDENTIALS_REGEXP = /^ *(?:[Bb][Aa][Ss][Ii][Cc]) +([A-Za-z0-9._~+/-]+=*) *$/
+const USER_PASS_REGEXP = /^([^:]*):(.*)$/
+const utf8Decoder = new TextDecoder()
+
+export type Auth = (req: HonoRequest) => { username: string; password: string } | undefined
```
Comment (yusukebe) [inline src/utils/basic-auth.ts:8]: I prefer making `req` as `Request` not `HonoRequest`. The utility function should be more general. So, `HonoRequest` that depends on Hono is not good.

**Comment by yusukebe** [comment]
Hi @sugar-cat7 

Thank you for the PR(this is your first PR for `honojs/hono`!). I've left a comment. Check it.

File: src/utils/basic-auth.ts (hunk 1 above)
Comment (sugar-cat7) [reply src/utils/basic-auth.ts]: Thank you for the review.
I have removed the dependency on HonoRequest.
https://github.com/honojs/hono/pull/3359/commits/989a0db600e2a3942b1e2257463b6fa7669a714f

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@sugar-cat7 

Thank you! I'm merging it.

**Pull request merged**